### PR TITLE
Getting System.Reflection.Metadata.Cil to build using dotnet

### PR DESCRIPTION
**Pull request by livarcocc**

### CODE REVIEW

**Comment by livarcocc** [comment]
@piotrpMSFT @weshaggard @sokket @schellap

File: dir.targets [hunk 1]
```diff
@@ -82,10 +93,10 @@
     <!-- Download latest nuget.exe -->
     <DownloadFile FileName="$(NuGetToolPath)"
                   Address="https://www.nuget.org/nuget.exe"
-                  Condition="!Exists('$(NuGetToolPath)') and '$(OsEnvironment)'=='Windows_NT'" />
+                  Condition="!Exists('$(NuGetToolPath)') and '$(OsEnvironment)'=='Windows_NT'" />
```
Comment (piotrpMSFT) [inline dir.targets]: why do we need these spaces?

File: dir.targets [hunk 2]
```diff
@@ -102,6 +113,16 @@
     <Exec Condition="'$(OsEnvironment)'=='Unix'"
           Command="find '$(RoslynPackageDir)tools' -name &quot;*.exe&quot; -exec chmod &quot;+x&quot; '{}' ';'" />
 
+    <!-- Download latest dotnet -->
+    <DownloadFile FileName="$(DotnetPackagePath)"
+                  Address="https://dotnetcli.blob.core.windows.net/dotnet/dev/Binaries/Latest/dotnet-win-x64.latest.zip"
```
Comment (piotrpMSFT) [inline dir.targets]: Can this be easily pulled out into a config?

File: dotnet-test.targets [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    <TestArguments Condition="'$(TestHost)'==''">$(XunitArguments)</TestArguments>
+
+    <TestCommandLine>$(TestProgram) $(TestArguments) {XunitTraitOptions}</TestCommandLine>
+  </PropertyGroup>
+
+  <Target Name="CopyTestToTestDirectory" 
+          BeforeTargets="RunTestsForProject">
+    <Message Text="Running CopyTestToTestDirectory: $(TestPath)%(TestTargetFramework.Folder)" Importance="High" />
+
+    <Exec Command="$(DotnetExe) publish &quot;$(ProjectJson)&quot; -o &quot;$(TestPath)%(TestTargetFramework.Folder)&quot; -f dnxcore50 -r win7-x64" />
```
Comment (piotrpMSFT) [inline dotnet-test.targets]: We'll want to update this once xunit runner comes online

File: src/System.Reflection.Metadata.Cil/src/CilExceptionRegion.cs [hunk 4]
```diff
@@ -68,6 +70,7 @@ public string ToString(CilTypeProvider provider)
             }
         }
 
+        [CLSCompliant(false)]
```
Comment (piotrpMSFT) [inline src/System.Reflection.Metadata.Cil/src/CilExceptionRegion.cs]: just curious, why are these all not CLSCompliant?

File: src/System.Reflection.Metadata.Cil/src/Decoder/CilTypeProvider.cs [hunk 5]
```diff
@@ -20,13 +21,15 @@ public MetadataReader Reader
...
         }
 
+        [CLSCompliant(false)]
         public CilTypeProvider(MetadataReader reader)
         {
             _reader = reader;
         }
 
-#region Public APIs
+        #region Public APIs
```
Comment (piotrpMSFT) [inline src/System.Reflection.Metadata.Cil/src/Decoder/CilTypeProvider.cs]: regions are not typically spaced in

**Comment by piotrpMSFT** [comment]
LGTM overall. Some small issues with spacing. Also curious why we need the CLSCompliant attribute everywhere now.

File: dir.targets [hunk 6]
```diff
@@ -59,6 +59,17 @@
     </Task>
   </UsingTask>
 
+  <UsingTask TaskName="UnzipFile" TaskFactory="CodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v12.0.dll">
```
Comment (weshaggard) [inline dir.targets]: FYI - If you plan to use this in the .NET Core version of msbuild or on mono neither support the CodeTaskFactory.

We actually have ZipFileFromDirectory build task (https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/ZipFileCreateFromDirectory.cs#L13) perhaps we should add an Extract.

File: src/System.Reflection.Metadata.Cil/src/Decoder/CilTypeProvider.cs [hunk 7]
```diff
@@ -229,9 +237,10 @@ public CilType GetTypeFromReference(TypeReferenceHandle handle, bool? isValueTyp
             return type;
         }
 
+        [CLSCompliant(false)]
         public string GetParameterList(MethodSignature<CilType> signature, ParameterHandleCollection? parameters = null)
         {
-            ImmutableArray<CilType> types = signature.ParameterTypes;
+            var types = signature.ParameterTypes;
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/src/Decoder/CilTypeProvider.cs]: Why the change to var? Now without going to definition I do not know what type these are.

File: src/System.Reflection.Metadata.Cil/src/Visitor/ICilVisitor.cs [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 namespace System.Reflection.Metadata.Cil.Visitor
 {
+    [CLSCompliant(false)]
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/src/Visitor/ICilVisitor.cs]: Any reason you didn't must mark the entire assembly as CLSCompliant(false)? or disable the check?

File: src/System.Reflection.Metadata.Cil/src/project.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  "dependencies": {    
+    "System.IO.FileSystem": "4.0.0",
+    "System.Linq": "4.0.0",
+    "System.Reflection": "4.0.0",    
+    "System.Reflection.TypeExtensions": "4.0.0",        
+  },
+  "frameworks": {
+    "dnxcore50": {
+      "dependencies":{
+        "System.Reflection.Metadata": "1.1.0-alpha-00014"
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/src/project.json]: Why are these dependencies under dnxcore50 and the others aren't?

File: src/System.Reflection.Metadata.Cil/tests/project.json [hunk 10]
```diff
@@ -0,0 +1,23 @@
+{
+  "name":"System.Reflection.Metadata.Cil.Tests",
+    "dependencies": {
+        "Microsoft.NETCore.TestHost": "1.0.0-beta-23409",
+        "Microsoft.NETCore.Platforms": "1.0.1-beta-23409",
+        "Microsoft.NETCore.Runtime.CoreCLR": "1.0.1-beta-23409",
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/tests/project.json]: It isn't recommended to directly reference the CoreCLR runtime specifically. Instead just reference Microsoft.NETCore.Runtime.

File: src/System.Reflection.Metadata.Cil/tests/project.json [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        "System.Console": "4.0.0-beta-23516",
+        "System.IO.FileSystem": "4.0.0",
+        "src": "1.0.0",
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/tests/project.json]: LOL... I assume this is the project reference to the library itself? We should figure out if there is a better way to do that as it is not at all clear that it is a project reference vs a package reference.

File: src/System.Reflection.Metadata.Cil/tests/System.Reflection.Metadata.Cil.Tests.xproj [hunk 12]
```diff
@@ -0,0 +1,19 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+  <Import Project="$(VSToolsPath)\DNX\Microsoft.DNX.Props" Condition="'$(VSToolsPath)' != ''" />
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/tests/System.Reflection.Metadata.Cil.Tests.xproj]: I assume this introduces a new dependency for building this repo/project? It now requires the DNX to be installed with VS (I am assuming that isn't installed by default).

File: src/System.Reflection.Metadata.Cil/src/System.Reflection.Metadata.Cil.xproj [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
+    <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+  </PropertyGroup>
+  <Import Project="$(VSToolsPath)\DNX\Microsoft.DNX.Props" Condition="'$(VSToolsPath)' != ''" />
+  <PropertyGroup Label="Globals">
+    <ProjectGuid>d654cecb-11b4-4da5-b418-5ef387e7705d</ProjectGuid>
+    <RootNamespace>src</RootNamespace>
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/src/System.Reflection.Metadata.Cil.xproj]: Why is the RootNamespace src?

File: src/System.Reflection.Metadata.Cil/src/CilTypeReference.cs [hunk 14]
```diff
@@ -14,7 +14,6 @@ public struct CilTypeReference
         private CilReaders _readers;
         private string _name;
         private string _namespace;
-        private CilEntity _resolutionScope;
```
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/src/CilTypeReference.cs]: How did you find this dead code?

File: src/System.Reflection.Metadata.Cil/src/Decoder/CilTypeProvider.cs (hunk 7 above)
Comment (weshaggard) [inline src/System.Reflection.Metadata.Cil/src/Decoder/CilTypeProvider.cs]: See rule 10 in our coding guidelines https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md

File: dir.targets (hunk 2 above)
Comment (livarcocc) [inline dir.targets]: Why would we want to configure this? If we decide to change it, we can update the script directly. In which instances out one person use one versus another uri?

File: dotnet-test.targets (hunk 3 above)
Comment (livarcocc) [inline dotnet-test.targets]: Absolutely. I am eagerly waiting for that. How is the assembly load work coming along.

File: dir.targets (hunk 6 above)
Comment (livarcocc) [inline dir.targets]: I will log an issue. About the Factory, there is another custom task defined in this file: downloadfile. I just used the same template. The end goal is to have it working on .NET Core, but I will take care of that once we have the unzip task. Sounds good?

File: src/System.Reflection.Metadata.Cil/tests/project.json (hunk 10 above)
Comment (livarcocc) [inline src/System.Reflection.Metadata.Cil/tests/project.json]: The problem is that  the Microsoft.NETCore.Runtime has a dependency that conflicts with another one in the project, so I couldn't directly depend on it.

File: src/System.Reflection.Metadata.Cil/tests/project.json (hunk 11 above)
Comment (livarcocc) [inline src/System.Reflection.Metadata.Cil/tests/project.json]: I added the type to it:

{
          "type": "build",
          "version": "1.0.0"
        }

File: src/System.Reflection.Metadata.Cil/src/project.json (hunk 9 above)
Comment (livarcocc) [inline src/System.Reflection.Metadata.Cil/src/project.json]: Moved.

File: src/System.Reflection.Metadata.Cil/tests/System.Reflection.Metadata.Cil.Tests.xproj (hunk 12 above)
Comment (livarcocc) [inline src/System.Reflection.Metadata.Cil/tests/System.Reflection.Metadata.Cil.Tests.xproj]: This is required so that VS can understand the xproj files appropriately. I would say this is a dependency between xproj and VS.
This is a very good point and we should address it.

File: src/System.Reflection.Metadata.Cil/src/System.Reflection.Metadata.Cil.xproj (hunk 13 above)
Comment (livarcocc) [inline src/System.Reflection.Metadata.Cil/src/System.Reflection.Metadata.Cil.xproj]: This file was generated. Updated the namespace.
